src/aspire/utils/relion_interop.py

```diff
@@ -1,11 +1,91 @@
-class RlnOpticsGroup:
-    def __init__(self, df_row):
+import logging
+from collections import OrderedDict
+
+import numpy as np
+
+from aspire.storage import StarFile
+
+logger = logging.getLogger(__name__)
+
+
+# The metadata_fields dictionary below specifies default data types
+# of certain key fields used in the codebase,
+# which are originally read from Relion STAR files.
+relion_metadata_fields = {
+    "_rlnVoltage": float,
+    "_rlnDefocusU": float,
+    "_rlnDefocusV": float,
+    "_rlnDefocusAngle": float,
+    "_rlnSphericalAberration": float,
+    "_rlnDetectorPixelSize": float,
+    "_rlnCtfFigureOfMerit": float,
+    "_rlnMagnification": float,
+    "_rlnAmplitudeContrast": float,
+    "_rlnImageName": str,
+    "_rlnOriginalName": str,
+    "_rlnCtfImage": str,
+    "_rlnCoordinateX": float,
+    "_rlnCoordinateY": float,
+    "_rlnCoordinateZ": float,
+    "_rlnNormCorrection": float,
+    "_rlnMicrographName": str,
+    "_rlnGroupName": str,
+    "_rlnGroupNumber": str,
+    "_rlnOriginX": float,
+    "_rlnOriginY": float,
+    "_rlnAngleRot": float,
+    "_rlnAngleTilt": float,
+    "_rlnAnglePsi": float,
+    "_rlnClassNumber": int,
+    "_rlnLogLikeliContribution": float,
+    "_rlnRandomSubset": int,
+    "_rlnParticleName": str,
+    "_rlnOriginalParticleName": str,
+    "_rlnNrOfSignificantSamples": float,
+    "_rlnNrOfFrames": int,
+    "_rlnMaxValueProbDistribution": float,
+    "_rlnOpticsGroup": int,
+    "_rlnOpticsGroupName": str,
+}
+
+
+def df_to_relion_types(df):
+    # convert STAR file strings to data type for each field
+    # columns without a specified data type are read as dtype=object
```

Comment:
This looks like a single comment over two lines. Maybe add periods or CAPS or both.

Reply:
added both